flowredux/src/commonMain/kotlin/com/freeletics/flowredux/sideeffects/SideEffect.kt

```diff
@@ -63,15 +92,62 @@ internal class SideEffectBuilder<InputState : S, S, A>(
     fun build() = builder()
 }
 
-internal typealias GetState<S> = () -> S
+internal class ManagedSideEffect<InputState : S, S, A>(
+    private val builder: SideEffectBuilder<InputState, S, A>,
+    private val scope: CoroutineScope,
+    private val getState: GetState<S>,
+    private val stateChanges: SendChannel<ChangedState<S>>,
+) {
 
-internal fun <InputState : S, S, A> SideEffect<InputState, S, A>.produceStateGuarded(
-    actions: Flow<Action<A>>,
-    getState: GetState<S>,
-): Flow<ChangedState<S>> {
-    return produceState(actions, getState).map { guardWithIsInState(it, this) }
+    private var current: Current? = null
+
+    suspend fun sendStateChange(state: S) {
+        if (builder.isInState.check(state)) {
+            var current = current
+            if (current == null) {
+                val currentSideEffect = builder.build()
+                val currentJob = scope.launch {
+                    currentSideEffect.produceState(getState).collect {
+                        // the side effect is in state might be more specific than the one in the builder
+                        stateChanges.send(guardWithIsInState(it, currentSideEffect))
+                    }
+                }
+                current = Current(currentJob, currentSideEffect)
+                this.current = current
+            }
+
+            current.sideEffect.sendState(state)
+        }
+    }
+
+    suspend fun cancelIfNeeded(state: S) {
+        val current = current
+        if (current != null) {
+            if (!current.sideEffect.isInState.check(state)) {
+                current.job.cancel(StateChangeCancellationException())
+                current.job.join()
+                this.current = null
+            }
+        }
+    }
+
+    suspend fun sendAction(action: A, state: S) {
+        val current = current
+        if (current != null && current.sideEffect.isInState.check(state)) {
+            current.sideEffect.sendAction(action)
+        }
+    }
+
+    private inner class Current(
```

Comment:
Maybe the naming could be a bit more specific. Is it the "Current Sideeffects for current State"? "Current Sideffect only"?  


Maybe you could think of a bit more expressive name to make the code a bit more readable. Up to you.

Reply:
What do you think of `CurrentlyActiveSideEffect`?

Reply:
[1e386bd](https://github.com/freeletics/FlowRedux/pull/508/commits/1e386bd2fc94b4a22501b7f4e27890f8fe6bb085)